docs/kb/semgrep-ci/jenkins-diff-scans.md

```diff
@@ -0,0 +1,185 @@
+
+### Create the Jenkinsfile
+
+Add the following to a Jenkinsfile in the root of the repository. This code snippet uses Jenkins declarative syntax and runs Semgrep in Docker.
+
+```bash
+pipeline {
+  agent any
+  environment {
+    // Required for a Semgrep Cloud Platform-connected scan:
+    SEMGREP_APP_TOKEN = credentials('SEMGREP_APP_TOKEN')
+    // Set repo name to expected format
+    SEMGREP_REPO_NAME = env.GIT_URL.replaceFirst(/^https:\/\/github.com\/(.*)$/, '$1')
+  }
+  stages {
+    stage('semgrep-diff-scan') {
+      when {
+        branch "PR-*"
+      }
+      steps {
+        sh '''git fetch --no-tags --force --progress -- $GIT_URL +refs/heads/$CHANGE_TARGET:refs/remotes/origin/$CHANGE_TARGET
+              git checkout -b $CHANGE_TARGET origin/$CHANGE_TARGET
+              git checkout $GIT_BRANCH
+           '''
+        sh '''docker pull semgrep/semgrep && \
+            docker run \
+            -e SEMGREP_APP_TOKEN=$SEMGREP_APP_TOKEN \
+            -e SEMGREP_REPO_NAME=$SEMGREP_REPO_NAME \
+            -e SEMGREP_BASELINE_REF=$(git merge-base $GIT_BRANCH $CHANGE_TARGET) \
+            -e SEMGREP_PR_ID = "${env.CHANGE_ID}"
+            -v "$(pwd):$(pwd)" --workdir $(pwd) \
+            semgrep/semgrep semgrep ci '''
+      }
+    }
+    stage('semgrep-scan') {
+      when {
+        branch "main"
+      }
+      steps {
+        sh '''docker pull semgrep/semgrep && \
+            docker run \
+            -e SEMGREP_APP_TOKEN=$SEMGREP_APP_TOKEN \
+            -e SEMGREP_REPO_NAME=$SEMGREP_REPO_NAME \
+            -v "$(pwd):$(pwd)" --workdir $(pwd) \
+            semgrep/semgrep semgrep ci '''
+      }
+    }
+  }
+  post {
+    // Clean after build
+    always {
+      cleanWs()
+    }
+  }
+}
+```
+
+The Jenkinsfile defines two Semgrep stages, each of which is run for certain branches: a diff-aware scan for PR branches, and a full scan for the main branch. This diff-aware scan configuration uses a computed merge base, rather than setting the merge base to the default branch. This is more similar to how Semgrep performs in GitHub actions. Setting the SEMGREP_REPO_NAME and SEMGREP_PR_ID allows Semgrep to identify the connected project and related PR.
+
+In order to compute the merge-base, the pipeline performs some additional git commands to make the default branch available to git for computation, and cleans the workspace afterwards so those commands can always run successfully in a clean workspace.
+
+:::info
+Using a computed merge base is strongly recommended. If you instead set `SEMGREP_BASELINE_REF` to `main` or `master`, you may see spurious findings in diff-aware scans if the remote branch has been updated independently of the PR branch, or the branch may not be available locally unless you perform a `git fetch` or `git checkout` as shown in this example.
```

Comment:
```suggestion
Using a computed merge base is strongly recommended. If you set `SEMGREP_BASELINE_REF` to `main` or `master` instead, you may see spurious findings in diff-aware scans if the remote branch has been updated independently of the PR or MR branch, or the branch may not be available locally unless you perform a `git fetch` or `git checkout`.
```

I don't know if it makes sense to update the final sentence to read, "...or the branch may not be available locally unless you perform a `git fetch` or `git checkout` **manually**"

Reply:
Not correct to add MR here either, for the same reason.